### PR TITLE
Dvci 170 - Lab Result Verification

### PR DESCRIPTION
**Pull request by mickohanlon23**
## Overview
This PR allows our app to scan and display SHCs that contain Lab Results. Credentials of [type #laboratory](https://spec.smarthealth.cards/vocabulary/) that comply with the [Generic Laboratory Result Observation profile](https://build.fhir.org/ig/HL7/fhir-shc-vaccination-ig/StructureDefinition-shc-infectious-disease-laboratory-result-observation-dm.html), can now be scanned, verified and displayed back to the user.

## What Changed
- Added `LabCard` component that gets used in `HealthCardDisplay` when a VC of type https://smarthealth.cards#laboratory is scanned
- `LabCard` component displays the `code`, `effectiveDateTime`, `value[x]`, `referenceRange`, and `performer` of the Observations in the SHC. Helper functions were added to handle elements that could have multiple data types. `code` and `value[x]` draw display values from the value sets that live in `src/components/LabCard`.
- An error is thrown if both #laboratory and #immunization types are found in the VC's `type` array, as we decided it is against the Vaccination and Testing IG to allow both in one Bundle.
- `LabCard.test.js` was added to test UI rendering of the LabCard component.

## To Test
- Make sure lab data is displaying appropriately
   -  Here is a test lab QR code that has all Must Support fields populated:
   - image
   - See [DVCI-170](https://jira.mitre.org/browse/DVCI-170) for a link to a real SHC.
- Check that the fields being displayed by the `LabCard` correspond to what would be expected from a resource conforming to the [Generic Lab Results profile](https://build.fhir.org/ig/HL7/fhir-shc-vaccination-ig/StructureDefinition-shc-infectious-disease-laboratory-result-observation-dm.html).
- Test different edge cases and error scenarios using our [Lab Result QR Code examples ](https://confluence.mitre.org/display/DVCI/Lab+Result+QR+Codes+for+Testing+the+Web+Verifier) on Confluence.
- Make sure you are still able to scan vaccine cards, and that no functionality was broken.

### CODE REVIEW

File: src/components/LabCard/LabCard.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+          <Grid item xs={9} sm={10} className={styles.gridItem}>
+            <Typography>
+              {displayLabTestResult(labResult)}
+            </Typography>
+          </Grid>
+        </Grid>
+
+        <Grid container item className={styles.gridRow} spacing={0}>
+          <Grid item xs={3} sm={2} className={styles.gridLabel}>
+            <Typography>{t('healthcarddisplay.Ref. Range')}</Typography>
```
Comment (sarahmcdougall) [inline src/components/LabCard/LabCard.js:118]: We talked a little about this in the working session - should we only display the field names if they are populated? ex. if no reference range value is included in the card, then don't display "Ref. Range" at all

File: src/components/LabCard/styles.jsx [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  },
+  divider: {
+    display: 'flex',
+    alignItems: 'center',
+    justifyContent: 'center',
+    alignSelf: 'flex-end',
+    height: '19px',
+    width: '100%',
+  },
+  vaccinationRecordBox: {
```
Comment (sarahmcdougall) [inline src/components/LabCard/styles.jsx:77]: Maybe we should rename some of these objects since this styling is for lab records rather than vaccination records (ex. `labRecordBox`)

**Review by sarahmcdougall — CHANGES_REQUESTED** [review]
This looks really good! I tested with various test lab QRs for different scenarios, which helped me better understand the code. I left a few open-ended questions to think about when displaying and testing lab results, so let me know if you want to talk through any of my comments.

I tested out the mismatch patient demographic data scenario and that still works as expected, so we can keep that code out of the `try` block. 

I did try scanning an immunization card, then clicked on "Add another SHC for same person" and scanned a lab result card without running into any errors. I know that we cannot have immunization and observation resources in the same card, so would that rule also apply to a situation like this where are scanning multiple SHCs for the same patient, but some are for vaccinations and some are for labs?

File: src/components/QrDataProvider.js [hunk 3]
```diff
@@ -48,67 +48,71 @@ const reducer = (state, action) => {
...
+            && card.birthDate === demographicData[0].birthDate
+        );
+        newState.matchingDemographicData = matchingDemographicData;
+
+        // Skip validation if it is disabled in App.config.js
         if (!config.ENABLE_VALIDATION) {
           newState.validationStatus = null;
+        // Validation for lab results is not currently supported
+        } else if (types.includes('https://smarthealth.cards#laboratory')) {
+          newState.validationStatus = null;
```
[the remark this answers is not in the document]
Comment (mickohanlon23) [reply src/components/QrDataProvider.js:105]: That's what I was originally thinking, that they were two separate "cases" and that lab validation could be supported in the future. But I don't think we have immediately plans to add lab validation, so it's probably best convention to combine them.

File: src/components/LabCard/LabCard.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    return (`${date} ${d.toLocaleTimeString('en-US')} (${timezone})`);
+  }
+
+  const displayReferenceRange = (referenceRange) => {
+    if (referenceRange.text) return referenceRange.text;
+    const lowVal = referenceRange.low;
+    const highVal = referenceRange.high;
+    if (lowVal && highVal) return (`${lowVal.value} ${lowVal.unit || ''} - ${highVal.value} ${highVal.unit || ''}`);
+    if (lowVal) return (`$>= ${lowVal.value} ${lowVal.unit || ''}`);
+    if (highVal) return (`$<= ${highVal.value} ${highVal.unit || ''}`);
```
[the remark this answers is not in the document]
Comment (mickohanlon23) [reply src/components/LabCard/LabCard.js:71]: Good catch

File: public/locales/en/translation.json [hunk 5]
```diff
@@ -18,6 +18,10 @@
     "Lot": "Lot",
     "Date": "Date",
     "Vaccinator": "Vaccinator",
+    "Lab Test": "Lab Test",
+    "Lab Performer": "Lab Performer",
+    "Ref. Range": "Ref. Range",
```
[the remark this answers is not in the document]
Comment (mickohanlon23) [reply public/locales/en/translation.json:23]: I had originally abbreviated it because its length was causing it to carry over into where the value is displayed, and I didn't want to make the styling inconsistent with the rest of the fields. But I agree, the abbreviation might not be intuitive. I tried moving "range" to a new line, but it makes it look like it's own field, which also could cause confusion and doesn't look too nice stylistically. Open to suggestions on what you think the best handling for this is.

[the remark this answers is not in the document]
Comment (mickohanlon23) [reply public/locales/en/translation.json]: In https://github.com/dvci/shc-web-verifier/pull/77/commits/357c65de50ef272c444fc8b4feb3151c2d8d5404, I increased the `paddingLeft` of the `gridItem` style so that the "Reference Range" grid label would not overlap with the reference range values displayed. Let me know what you think.

File: src/components/QrDataProvider.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mickohanlon23) [reply src/components/QrDataProvider.js]: Fixed: https://github.com/dvci/shc-web-verifier/pull/77/commits/ec4238830cf9588c321a2b3966d965712972517a

File: src/components/LabCard/styles.jsx (hunk 2 above)
Comment (mickohanlon23) [reply src/components/LabCard/styles.jsx]: I renamed all of the `vaccination...` type styles to `lab...` in https://github.com/dvci/shc-web-verifier/pull/77/commits/08573dca3ce250cd16b5d9356d7a6991245f1442

File: public/locales/en/translation.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (mickohanlon23) [reply public/locales/en/translation.json]: This was slightly modified in https://github.com/dvci/shc-web-verifier/pull/77/commits/fd1aa80c6f0aa89dcc28d2fa624c3df354ecc673 so that all the grid labels are aligned.

**Comment by mickohanlon23** [comment]
> This looks really good! I tested with various test lab QRs for different scenarios, which helped me better understand the code. I left a few open-ended questions to think about when displaying and testing lab results, so let me know if you want to talk through any of my comments.
> 
> I tested out the mismatch patient demographic data scenario and that still works as expected, so we can keep that code out of the `try` block.
> 
> I did try scanning an immunization card, then clicked on "Add another SHC for same person" and scanned a lab result card without running into any errors. I know that we cannot have immunization and observation resources in the same card, so would that rule also apply to a situation like this where are scanning multiple SHCs for the same patient, but some are for vaccinations and some are for labs?

I don't necessarily think so, but I guess that's up for us to decide. I think we didn't want to support displaying both resources from the same SHC, since this would technically make that card noncompliant with the vaccination and testing FHIR IG, so we're assuming that those type of cards won't exist. But there technically isn’t anything wrong with someone displaying their immunization card and then their lab card, so I don't see a strong reason why we would need to throw an error here.

A potential use case could be that someone shows their card with only one vaccine that doesn’t make them "fully vaccinated". We display "Cannot determine vaccination status", but then they also go on to scan their lab card showing a recent negative test, proving that they've met criteria to enter the venue in question.

**Review by sarahmcdougall — CHANGES_REQUESTED** [review]
Thanks for adding those unit tests and styling changes! I like the decision to only include the reference range field if the card contains a reference range, and I like the new ordering of the fields on the SMART Health Card.

When scanning example QRs, I realized the DOB styling doesn't match the changes from [Zara's PR](https://github.com/dvci/shc-web-verifier/pull/70). We should use the same styling for both the vaccination and immunization cards - should be an easy fix :)

Also, I noticed that this is what gets displayed when the card does not contain the #laboratory type:

Screen Shot 2022-07-29 at 10 17 24 AM

Is this the expected display? It feels odd that the card doesn't have the same width as the "Scan QR" button.

**Comment by mickohanlon23** [comment]
> Thanks for adding those unit tests and styling changes! I like the decision to only include the reference range field if the card contains a reference range, and I like the new ordering of the fields on the SMART Health Card.
> 
> When scanning example QRs, I realized the DOB styling doesn't match the changes from [Zara's PR](https://github.com/dvci/shc-web-verifier/pull/70). We should use the same styling for both the vaccination and immunization cards - should be an easy fix :)
> 
> Also, I noticed that this is what gets displayed when the card does not contain the #laboratory type:
> 
> Screen Shot 2022-07-29 at 10 17 24 AM
> 
> Is this the expected display? It feels odd that the card doesn't have the same width as the "Scan QR" button.

@sarahmcdougall I fixed the DOB styling in https://github.com/dvci/shc-web-verifier/pull/77/commits/4c7d656ce534fcff6d139d2b4a2ddd8039da7a01.

The shrinking card width is an existing UI quirk independent of the lab results display, so I created [DVCI-332](https://jira.mitre.org/browse/DVCI-332) to address it.

**Review by sarahmcdougall — APPROVED** [review]

File: src/components/LabCard/LabCard.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            </Grid>
+          </Grid>
+        )}
+
+      </Grid>
+    </Box>
+  );
+
+  return (
+    <Box display="flex" className={styles.healthCard}>
```
Comment (nkarip) [inline src/components/LabCard/LabCard.js:147]: This wrapping portion and styling is shared between lab and immunizations and should be consolidated in a base component.

**Review by nkarip — CHANGES_REQUESTED** [review]
There is a lot of elements and styling in LabCard and VaccineCard that are shared and should be extracted out to a re-usable component, especially since the styling can get complicated when mobile resizing is modified. We don't want to have to maintain duplicate style files for these.

File: src/components/LabCard/LabCard.js (hunk 6 above)
Comment (mickohanlon23) [reply src/components/LabCard/LabCard.js]: Good call, I agree. I made a `HealthCard` component which returns the Card itself. This contains the patient information, and Issuer. It wraps either the `VaccineCard` or the `LabCard` depending on which `credentialType` is presented.

**Review by nkarip — APPROVED** [review]
